### PR TITLE
add google oauth

### PR DESCRIPTION
**Pull request by dawntomm**

### CODE REVIEW

**Comment by byliuyang** [comment]
Hi @dawntomm , please fill in pull request description following: https://github.com/google/eng-practices/blob/master/review/developer/cl-descriptions.md. Please also address all codeclimates' comments.

File: backend/cmd/start.go [hunk 1]
```diff
@@ -20,12 +20,21 @@ type FacebookConfig struct {
 	RedirectURI  string
 }
 
+// GoogleConfig contains Facebook OAuth config
```
Comment (byliuyang) [inline backend/cmd/start.go]: I wonder why GoogleConfig contains Facebook info?

File: backend/cmd/start.go [hunk 2]
```diff
@@ -20,12 +20,21 @@ type FacebookConfig struct {
...
+type GoogleConfig struct {
+	ClientID     string
+	ClientSecret string
+	RedirectURI  string
```
Comment (byliuyang) [inline backend/cmd/start.go]: Are you using RedirectURI in the code?

File: backend/go.mod [hunk 3]
```diff
@@ -13,6 +13,7 @@ require (
 	github.com/spf13/pflag v1.0.5 // indirect
 	golang.org/x/net v0.0.0-20191116160921-f9c825593386 // indirect
 	golang.org/x/sys v0.0.0-20191118133127-cf1e2d577169 // indirect
+	golang.org/x/tools v0.0.0-20191203043002-5ae4576c3a41 // indirect
```
Comment (byliuyang) [inline backend/go.mod]: I wonder why is this added?

File: backend/go.sum [hunk 4]
```diff
@@ -143,6 +143,7 @@ golang.org/x/net v0.0.0-20190213061140-3a22650c66bd/go.mod h1:mL1N/T3taQHkDXs73r
 golang.org/x/net v0.0.0-20190311183353-d8887717615a/go.mod h1:t9HGtf8HONx5eT2rtn7q6eTqICYqUVnKs3thJo3Qplg=
 golang.org/x/net v0.0.0-20190404232315-eb5bcb51f2a3/go.mod h1:t9HGtf8HONx5eT2rtn7q6eTqICYqUVnKs3thJo3Qplg=
 golang.org/x/net v0.0.0-20190603091049-60506f45cf65/go.mod h1:HSz+uSET+XFnRR8LxR5pz3Of3rY3CfYBVs4xY44aLks=
+golang.org/x/net v0.0.0-20190620200207-3b0461eec859/go.mod h1:z5CRVTTTmAJ677TzLLGU+0bjPO0LkuOLi4/5GtJWs/s=
```
Comment (byliuyang) [inline backend/go.sum]: Is this necessary?

File: backend/go.sum [hunk 5]
```diff
@@ -177,6 +178,9 @@ golang.org/x/tools v0.0.0-20190524140312-2c0ae7006135/go.mod h1:RgjU9mgBXZiqYHBn
 golang.org/x/tools v0.0.0-20190606124116-d0a3d012864b/go.mod h1:/rFqwRUd4F7ZHNgwSSTFct+R/Kf4OFW1sUzUTQQTgfc=
 golang.org/x/tools v0.0.0-20190624180213-70d37148ca0c h1:KfpJVdWhuRqNk4XVXzjXf2KAV4TBEP77SYdFGjeGuIE=
 golang.org/x/tools v0.0.0-20190624180213-70d37148ca0c/go.mod h1:/rFqwRUd4F7ZHNgwSSTFct+R/Kf4OFW1sUzUTQQTgfc=
+golang.org/x/tools v0.0.0-20191203043002-5ae4576c3a41 h1:bFvRUNBUPLnnjjl4tac0PRqStjfXeTZwpQ49o7F2DLk=
+golang.org/x/tools v0.0.0-20191203043002-5ae4576c3a41/go.mod h1:b+2E5dAYhXwXZwtnZ6UAqBI28+e2cm9otk0dWdXHAEo=
+golang.org/x/xerrors v0.0.0-20190717185122-a985d3407aa7/go.mod h1:I/5z698sn9Ka8TeJc9MKroUUfqBBauWjQqLJ2OPfmY0=
```
Comment (byliuyang) [inline backend/go.sum]: Are those necessary?

File: frontend/src/service/StaticConfigDecision.service.ts [hunk 6]
```diff
@@ -2,10 +2,10 @@ import { IFeatureDecisionService } from './FeatureDecision.service';
 
 export class StaticConfigDecisionService implements IFeatureDecisionService {
   includeGithubSignButton(): boolean {
-    return false;
+    return true;
```
Comment (byliuyang) [inline frontend/src/service/StaticConfigDecision.service.ts]: Not now.

File: backend/app/adapter/google/account.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+	"short/app/entity"
+	"short/app/usecase/service"
+	"strings"
+)
+
+var _ service.SSOAccount = (*Account)(nil)
+
+// Account accesses user's account data through FB Graph API.
+type Account struct {
+}
```
[the remark this answers is not in the document]
Comment (dawntomm) [reply backend/app/adapter/google/account.go]: changed accordingly

File: backend/app/adapter/google/account.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+// GetSingleSignOnUser retrieves user's email and name from Facebook API.
+func (g Account) GetSingleSignOnUser(accessToken string) (entity.SSOUser, error) {
+	type response struct {
+		Email string `json:"email"`
+		Name  string `json:"name"`
+	}
+
+	form := url.Values{}
+
+	req, err := http.NewRequest("GET", "https://www.googleapis.com/drive/v2/files", strings.NewReader(form.Encode()))
```
[the remark this answers is not in the document]
Comment (dawntomm) [reply backend/app/adapter/google/account.go]: according to https://developers.google.com/identity/protocols/OAuth2WebServer, google drive api is the correct url

[the remark this answers is not in the document]
Comment (byliuyang) [reply backend/app/adapter/google/account.go]: @dawntomm Please checkout this: https://developers.google.com/identity/protocols/googlescopes#google_sign-in

File: backend/app/adapter/google/identityprovider.go [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+	"fmt"
+	"net/http"
+	"net/url"
+	"short/app/usecase/service"
+
+	"github.com/byliuyang/app/fw"
+)
+
+const (
+	authorizationAPI     = "https://accounts.google.com/o/oauth2/v2/auth?scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.metadata.readonly&access_type=offline&include_granted_scopes=true&state=state_parameter_passthrough_value&redirect_uri=http%3A%2F%2Flocalhost%2Foauth%2Fgoogle%2Fsign-in%2Fcallback&response_type=code"
```
[the remark this answers is not in the document]
Comment (dawntomm) [reply backend/app/adapter/google/identityprovider.go]: removed.

[the remark this answers is not in the document]
Comment (dawntomm) [reply backend/app/adapter/google/identityprovider.go]: can you give me a  clear text example?

File: backend/go.mod (hunk 3 above)
Comment (dawntomm) [reply backend/go.mod]: these are auto generated

File: backend/go.sum (hunk 4 above)
Comment (dawntomm) [reply backend/go.sum]: these are auto generated

File: frontend/src/service/StaticConfigDecision.service.ts (hunk 6 above)
Comment (dawntomm) [reply frontend/src/service/StaticConfigDecision.service.ts]: ok

File: backend/app/adapter/google/identityprovider.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (byliuyang) [reply backend/app/adapter/google/identityprovider.go]: @dawntomm localhost means that Google sign in is redirecting to your local machine. However, when we deploy the app to staging or production environment, the redirect URL should then become `https://api-staging.short-d.com`. Facebook `IdentityProvider` handles this very well with environmental variable.

```go
// NewIdentityProvider initializes Facebook OAuth service.
func NewIdentityProvider(
	http fw.HTTPRequest,
	clientID string,
	clientSecret string,
	redirectURI string,
) IdentityProvider {
	return IdentityProvider{
		clientID:     clientID,
		clientSecret: clientSecret,
		http:         http,
		redirectURI:  url.QueryEscape(redirectURI),
	}
}
```

File: backend/go.mod (hunk 3 above)
Comment (byliuyang) [reply backend/go.mod]: @dawntomm May I ask what command did you use to automatically generate them?

File: backend/go.sum (hunk 4 above)
Comment (byliuyang) [reply backend/go.sum]: @dawntomm May I ask what command did you use to automatically generate them?

File: backend/app/adapter/google/account.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (dawntomm) [reply backend/app/adapter/google/account.go]: i believe scope is different from api

File: backend/app/adapter/google/identityprovider.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (dawntomm) [reply backend/app/adapter/google/identityprovider.go]: changed accordingly
